apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java

```diff
@@ -53,8 +56,12 @@
     /**
      * A mark captures the timing of a significant event during the lifetime of a transaction. Marks are organized into groups and can be set by the user or the agent.
      */
-    private final Map<String, Object> marks = new HashMap<>();
+    private final Map<String, Object> marks = new ConcurrentHashMap<>();
     private final SpanCount spanCount = new SpanCount();
+    /**
+     * Current span created for this transaction or null if there is no span at the moment. If there are nested spans this should contain the deepest.
+     */
+    private volatile Span currentSpan;
```

Comment:
I'm not sure that approach works. For example, if you are starting two spans in parallel there could be a race condition so that one span is wrongly declared as the parent of another span when they should, in fact, be both direct children of the transaction. I'd rather remove the current span here and add `Span#createSpan` (or `Span.createChildSpan`?) as the only way of creating nested spans. I'd even consider removing `ElasticApmTracer#createSpan`, but I'm not certain about that yet.

Reply:
Agree. Also though about that.